File: src/lava/magma/runtime/message_infrastructure/test/cpptest/test_grpc_delivery.cc
// Copyright (C) 2022 Intel Corporation
// SPDX-License-Identifier: BSD-3-Clause
// See: https://spdx.org/licenses/

#include <message_infrastructure/csrc/core/channel_factory.h>
#include <message_infrastructure/csrc/core/multiprocessing.h>
#include <message_infrastructure/csrc/channel/grpc/grpc_channel.h>
#include <message_infrastructure/csrc/core/message_infrastructure_logging.h>
#include <message_infrastructure/csrc/core/utils.h>
#include <gtest/gtest.h>
#include <iostream>

namespace message_infrastructure {

static void stop_fn() {
  // exit(0);
}

GrpcMetaDataPtr MetaData2GrpcMetaData(MetaDataPtr metadata) {
  GrpcMetaDataPtr grpcdata = std::make_shared<GrpcMetaData>();
  grpcdata->set_nd(metadata->nd);
  grpcdata->set_type(metadata->type);
  grpcdata->set_elsize(metadata->elsize);
  grpcdata->set_total_size(metadata->total_size);
  // char* data = reinterpret_cast<char*>(metadata->mdata);
  for (int i = 0; i < metadata->nd; i++) {
    grpcdata->add_dims(metadata->dims[i]);
    grpcdata->add_strides(metadata->strides[i]);
  }
  grpcdata->set_value(metadata->mdata, metadata->elsize*metadata->total_size);
  return grpcdata;
}

void grpc_target_fn1(
  int loop,
  AbstractChannelPtr mp_to_a1,
  AbstractChannelPtr a1_to_mp,
  AbstractChannelPtr a1_to_a2,
  AbstractChannelPtr a2_to_a1,
  AbstractActor* actor_ptr) {
    actor_ptr->SetStopFn(stop_fn);
    auto from_mp = mp_to_a1->GetRecvPort();
    auto to_mp = a1_to_mp->GetSendPort();
    auto to_a2   = a1_to_a2->GetSendPort();
    auto from_a2 = a2_to_a1->GetRecvPort();
    from_mp->Start();
    to_mp->Start();
    to_a2->Start();
    from_a2->Start();
    LAVA_DUMP(1, "grpc actor1, loop: %d\n", loop);
    while ((loop--)&&!actor_ptr->GetStatus()) {
      LAVA_DUMP(LOG_UTTEST, "grpc actor1 waitting\n");
      MetaDataPtr data = from_mp->Recv();
      LAVA_DUMP(LOG_UTTEST, "grpc actor1 recviced\n");
      (*reinterpret_cast<int64_t*>(data->mdata))++;
      to_a2->Send(MetaData2GrpcMetaData(data));
      free(reinterpret_cast<char*>(data->mdata));
      data = from_a2->Recv();
      (*reinterpret_cast<int64_t*>(data->mdata))++;
      to_mp->Send(MetaData2GrpcMetaData(data));
      free(reinterpret_cast<char*>(data->mdata));
    }
    from_mp->Join();
    to_mp->Join();
    to_a2->Join();
    from_a2->Join();
    while (!actor_ptr->GetStatus()) {
      helper::Sleep();
    }
  }

void grpc_target_fn2(
  int loop,
  AbstractChannelPtr a1_to_a2,
  AbstractChannelPtr a2_to_a1,
  AbstractActor* actor_ptr) {
    actor_ptr->SetStopFn(stop_fn);
    auto to_a1 = a2_to_a1->GetSendPort();
    auto from_a1 = a1_to_a2->GetRecvPort();
    from_a1->Start();
    to_a1->Start();
    LAVA_DUMP(LOG_UTTEST, "grpc actor2, loop: %d\n", loop);
    while ((loop--)&&!actor_ptr->GetStatus()) {
      LAVA_DUMP(LOG_UTTEST, "grpc actor2 waitting\n");
      MetaDataPtr data = from_a1->Recv();
      LAVA_DUMP(LOG_UTTEST, "grpc actor2 recviced\n");
      (*reinterpret_cast<int64_t*>(data->mdata))++;
      to_a1->Send(MetaData2GrpcMetaData(data));
      free(reinterpret_cast<char*>(data->mdata));
    }
    from_a1->Join();
    to_a1->Join();
    while (!actor_ptr->GetStatus()) {
      helper::Sleep();
    }
  }

TEST(TestGRPCChannel, GRPCLoop) {
  MultiProcessing mp;
  int loop = 1000;
  AbstractChannelPtr mp_to_a1 = GetChannelFactory().GetDefRPCChannel(
    "mp_to_a1", "mp_to_a1", 6);
  AbstractChannelPtr a1_to_mp = GetChannelFactory().GetDefRPCChannel(
    "a1_to_mp", "a1_to_mp", 6);
  AbstractChannelPtr a1_to_a2 = GetChannelFactory().GetDefRPCChannel(
    "a1_to_a2", "a1_to_a2", 6);
  AbstractChannelPtr a2_to_a1 = GetChannelFactory().GetDefRPCChannel(
    "a2_to_a1", "a2_to_a1", 6);
  auto target_fn_a1 = std::bind(&grpc_target_fn1,
                                loop,
                                mp_to_a1,
                                a1_to_mp,
                                a1_to_a2,
                                a2_to_a1,
                                std::placeholders::_1);
  auto target_fn_a2 = std::bind(&grpc_target_fn2,
                                loop,
                                a1_to_a2,
                                a2_to_a1,
                                std::placeholders::_1);
  int actor1 = mp.BuildActor(target_fn_a1);
  int actor2 = mp.BuildActor(target_fn_a2);
  auto to_a1 = mp_to_a1->GetSendPort();
  to_a1->Start();
  auto from_a1 = a1_to_mp->GetRecvPort();
  from_a1->Start();
  int64_t array_[10000] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 0};
  std::fill(array_ + 10, array_ + 10000, 1);

  MetaDataPtr metadata = std::make_shared<MetaData>();
  int64_t* array = reinterpret_cast<int64_t*>(array_);
  int64_t dims[] = {10000, 0, 0, 0, 0};
  int64_t nd = 1;

  GetMetadata(metadata, array, nd, METADATA_TYPES::LONG, dims);
  int expect_result = 1 + loop * 3;
  const clock_t start_time = std::clock();
  while (loop--) {
    LAVA_DUMP(LOG_UTTEST, "wait for response, remain loop: %d\n", loop);
    to_a1->Send(MetaData2GrpcMetaData(metadata));
    metadata = from_a1->Recv();
    LAVA_DUMP(LOG_UTTEST, "metadata:\n");
    LAVA_DUMP(LOG_UTTEST, "nd: %ld\n", metadata->nd);
    LAVA_DUMP(LOG_UTTEST, "type: %ld\n", metadata->type);
    LAVA_DUMP(LOG_UTTEST, "elsize: %ld\n", metadata->elsize);
    LAVA_DUMP(LOG_UTTEST, "total_size: %ld\n", metadata->total_size);
    LAVA_DUMP(LOG_UTTEST, "dims: {%ld, %ld, %ld, %ld, %ld}\n",
              metadata->dims[0], metadata->dims[1], metadata->dims[2],
              metadata->dims[3], metadata->dims[4]);
    LAVA_DUMP(LOG_UTTEST, "strides: {%ld, %ld, %ld, %ld, %ld}\n",
              metadata->strides[0], metadata->strides[1], metadata->strides[2],
              metadata->strides[3], metadata->strides[4]);
    LAVA_DUMP(LOG_UTTEST, "grpc mdata: %p, grpc *mdata: %ld\n", metadata->mdata,
              *reinterpret_cast<int64_t*>(metadata->mdata));
    free(reinterpret_cast<char*>(metadata->mdata));
  }
  const clock_t end_time = std::clock();
  to_a1->Join();
  from_a1->Join();
  int64_t result = *reinterpret_cast<int64_t*>(metadata->mdata);
  LAVA_DUMP(LOG_UTTEST, "grpc result =%ld\n", result);
  mp.Stop(true);
  if (result != expect_result) {
    LAVA_DUMP(LOG_UTTEST, "expect_result: %d\n", expect_result);
    printf("result: %ld\n", result);
    LAVA_DUMP(LOG_UTTEST, "result: %ld\n", result);
    LAVA_LOG_ERR("result != expect_result\n");
    throw;
  }
  std::printf("grpc cpp loop timedelta: %f\n",
           ((end_time - start_time)/static_cast<double>(CLOCKS_PER_SEC)));
  LAVA_DUMP(LOG_UTTEST, "exit\n");
}

}  // namespace message_infrastructure

File: src/lava/magma/runtime/message_infrastructure/test/test_grpc_delivery.py
# Copyright (C) 2022 Intel Corporation
# SPDX-License-Identifier: BSD-3-Clause
# See: https://spdx.org/licenses/

import numpy as np
import unittest
from functools import partial
import time
from datetime import datetime

from message_infrastructure.multiprocessing import MultiProcessing

from message_infrastructure import (
    ChannelBackend,
    Channel,
    SendPort,
    RecvPort,
    SupportGRPCChannel,
    ChannelQueueSize
)


class Builder:
    def build(self):
        pass


def prepare_data():
    return np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 0])


def actor_stop(name):
    pass


def target_fn1(*args, **kwargs):
    actor = args[0]
    actor.set_stop_fn(partial(actor_stop, "send"))
    loop = kwargs.pop("loop")
    a1_to_a2 = kwargs.pop("a1_to_a2")
    a2_to_a1 = kwargs.pop("a2_to_a1")
    mp_to_a1 = kwargs.pop("mp_to_a1")
    a1_to_mp = kwargs.pop("a1_to_mp")
    from_mp = mp_to_a1.dst_port
    to_mp = a1_to_mp.src_port
    to_a2 = a1_to_a2.src_port
    from_a2 = a2_to_a1.dst_port
    from_mp.start()
    to_mp.start()
    from_a2.start()
    to_a2.start()
    while loop:
        data = from_mp.recv()
        data[0] = data[0] + 1
        to_a2.send(data)
        data = from_a2.recv()
        data[0] = data[0] + 1
        to_mp.send(data)
        loop -= 1
    from_mp.join()
    to_mp.join()
    from_a2.join()
    to_a2.join()
    actor.status_stopped()


def target_fn2(*args, **kwargs):
    actor = args[0]
    actor.set_stop_fn(partial(actor_stop, "send"))
    loop = kwargs.pop("loop")
    a1_to_a2 = kwargs.pop("a1_to_a2")
    a2_to_a1 = kwargs.pop("a2_to_a1")
    from_a1 = a1_to_a2.dst_port
    to_a1 = a2_to_a1.src_port
    from_a1.start()
    to_a1.start()
    while loop:
        data = from_a1.recv()
        data[0] = data[0] + 1
        to_a1.send(data)
        loop -= 1

    from_a1.join()
    to_a1.join()
    actor.status_stopped()


class TestGrpcDelivery(unittest.TestCase):
    @unittest.skipIf(not SupportGRPCChannel, "Not support grpc channel.")
    def test_grpcchannel(self):
        from message_infrastructure import GetRPCChannel
        mp = MultiProcessing()
        mp.start()
        loop = 10000
        a1_to_a2 = GetRPCChannel(
            '127.13.2.15',
            8003,
            'a1_to_a2',
            'a1_to_a2', 8)
        a2_to_a1 = GetRPCChannel(
            '127.13.2.16',
            8004,
            'a2_to_a1',
            'a2_to_a1', 8)
        mp_to_a1 = GetRPCChannel(
            '127.13.2.17',
            8005,
            'mp_to_a1',
            'mp_to_a1', 8)
        a1_to_mp = GetRPCChannel(
            '127.13.2.18',
            8006,
            'a1_to_mp',
            'a1_to_mp', 8)

        recv_port_fn = partial(target_fn1,
                               loop=loop,
                               mp_to_a1=mp_to_a1,
                               a1_to_mp=a1_to_mp,
                               a1_to_a2=a1_to_a2,
                               a2_to_a1=a2_to_a1)
        send_port_fn = partial(target_fn2,
                               loop=loop,
                               a1_to_a2=a1_to_a2,
                               a2_to_a1=a2_to_a1)
        builder1 = Builder()
        builder2 = Builder()
        mp.build_actor(recv_port_fn, builder1)
        mp.build_actor(send_port_fn, builder2)
        to_a1 = mp_to_a1.src_port
        from_a1 = a1_to_mp.dst_port
        to_a1.start()
        from_a1.start()
        data = prepare_data()
        expect_result = prepare_data()
        expect_result[0] = (1 + 3 * loop)
        loop_start_time = datetime.now()
        while loop:
            to_a1.send(data)
            data = from_a1.recv()
            loop -= 1
        loop_end_time = datetime.now()
        from_a1.join()
        to_a1.join()
        mp.stop(True)
        if not np.array_equal(expect_result, data):
            print("expect: ", expect_result)
            print("result: ", data)
            raise AssertionError()
        print("cpp_grpc_loop_with_cpp_multiprocess timedelta =",
              loop_end_time - loop_start_time)


if __name__ == "__main__":
    unittest.main()
